clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -267,15 +267,19 @@ public void handshake() throws IOException {
         } catch (IOException e) {
             maybeThrowSslAuthenticationException();
 
-            // this exception could be due to a write. If there is data available to unwrap,
-            // process the data so that any SSL handshake exceptions are reported
-            if (handshakeStatus == HandshakeStatus.NEED_UNWRAP && netReadBuffer.position() > 0) {
+            // This exception could be due to a write. If there is data available to unwrap in the buffer, or data available
+            // in the socket channel to read and unwrap, process the data so that any SSL handshake exceptions are reported.
+            do {
                 try {
                     handshakeUnwrap(false);
+                    if (key.isReadable())
+                        read = readFromSocketChannel();
                 } catch (SSLException e1) {
                     maybeProcessHandshakeFailure(e1, false, e);
                 }
-            }
+
+            } while (netReadBuffer.position() > 0 && read > 0);
```

Comment:
Discussed this offline with @rajinisivaram and we think we think the first condition can be removed.

Reply:
@ijuma Thanks for the review, updated.